core/lib/zksync_core/src/l1_gas_price/gas_adjuster/mod.rs

```diff
@@ -127,24 +130,39 @@ impl<E: EthInterface> GasAdjuster<E> {
 impl<E: EthInterface> L1GasPriceProvider for GasAdjuster<E> {
     /// Returns the sum of base and priority fee, in wei, not considering time in mempool.
     /// Can be used to get an estimate of current gas price.
-    fn estimate_effective_gas_price(&self) -> u64 {
+    fn estimate_effective_gas_price(&self) -> U256 {
         if let Some(price) = self.config.internal_enforced_l1_gas_price {
-            return price;
+            return U256::from(price);
         }
 
         let effective_gas_price = self.get_base_fee(0) + self.get_priority_fee();
 
         let calculated_price =
             (self.config.internal_l1_pricing_multiplier * effective_gas_price as f64) as u64;
 
-        let conversion_rate = self.native_token_fetcher.conversion_rate().unwrap_or(1);
-
-        self.bound_gas_price(calculated_price) * conversion_rate
+        // To convert from BigDecimal to U256:
+        // 1. Round the value to the nearest integer.
+        // 2. Convert the rounded value to a string.
+        // 3. Convert the string to U256.
+        // - In case of conversion failure, assume it's overflow and return max value.
+        // -- TODO: if the value is greater or equal to U256::max_value(),
+        // --       the return value might also overflow.
+        let conversion_rate = U256::from_dec_str(
+            &self
+                .native_token_fetcher
+                .conversion_rate()
+                .unwrap_or(BigDecimal::one())
+                .round(0)
+                .to_string(),
+        )
+        .unwrap_or(U256::max_value());
```

Comment:
Cumbersome conversion, should also debate with zk if it's actually correct to assume that a casting error is caused by an overflow and thereby assing max possible value.